substrate/frame/scheduler/src/lib.rs

```diff
@@ -1157,24 +1190,51 @@ impl<T: Config> Pallet<T> {
 			return
 		}
 
-		let mut incomplete_since = now + One::one();
-		let mut when = IncompleteSince::<T>::take().unwrap_or(now);
-		let mut executed = 0;
+		let queue = Queue::<T>::get();
+		let end_index = match queue.iter().position(|&x| x >= now) {
+			Some(end_index) => end_index.saturating_add(1),
+			None => {
+				return;
+			},
+		};
+		if end_index == 0 {
+			return;
+		}
 
-		let max_items = T::MaxScheduledPerBlock::get();
-		let mut count_down = max;
-		let service_agenda_base_weight = T::WeightInfo::service_agenda_base(max_items);
-		while count_down > 0 && when <= now && weight.can_consume(service_agenda_base_weight) {
-			if !Self::service_agenda(weight, &mut executed, now, when, u32::max_value()) {
-				incomplete_since = incomplete_since.min(when);
+		let mut index = 0;
+		let queue_len = queue.len();
+		for when in queue.iter().skip(index) {
+			if *when < now.saturating_sub(T::MaxStaleTaskAge::get()) {
+				Agenda::<T>::remove(*when);
```

Comment:
Why do we need this `MaxStaleTaskAge`? I think it should maybe rather be checked on inserting into an agenda instead of when servicing it, since in the service step we have nothing to lose from executing it.  
But am not sure if we need it at all.